pkg/mount/listener_linux.go

```diff
@@ -0,0 +1,229 @@
+	prevMountMap map[string]MountInfo
+	closeCh      chan struct{}
+	mutex        sync.Mutex
+	eventMap     map[string]*Event
+	keys         []string
+	waitCh       chan struct{}
+}
+
+func (listener *Listener) Close() error {
+	if atomic.AddInt32(&listener.isClosed, 1) == 1 {
+		listener.mountFile.Close()
+		close(listener.closeCh)
+		return syscall.Close(listener.epollFd)
+	}
+	return nil
+}
+
+func (listener *Listener) get() (string, *Event, <-chan struct{}, error) {
+	listener.mutex.Lock()
+	defer listener.mutex.Unlock()
+
+	if len(listener.eventMap) == 0 && len(listener.keys) > 0 {
+		listener.keys = []string{}
+	}
+
+	var event *Event
+	var found bool
+	for !found {
+		if len(listener.keys) == 0 {
+			break
+		}
+
+		key := listener.keys[0]
+		listener.keys = listener.keys[1:]
+		if event, found = listener.eventMap[key]; found {
+			delete(listener.eventMap, key)
+			return key, event, nil, nil
+		}
+	}
+
+	// As no event found, wait for event to be set.
+	waitCh := listener.waitCh
+	if waitCh == nil {
+		waitCh = make(chan struct{})
+		listener.waitCh = waitCh
+	}
+	return "", nil, waitCh, nil
+}
+
+func (listener *Listener) Get(ctx context.Context) (string, *Event, error) {
+	for {
+		key, event, waitCh, err := listener.get()
+		switch {
+		case err != nil:
+			return "", nil, err
+		case event != nil && key != "":
+			switch event.Type() {
+			case Attached, Modified:
+				listener.prevMountMap[key] = event.MountInfo()
+			case Detached:
+				delete(listener.prevMountMap, key)
+			}
+			return key, event, nil
+		}
+
+		select {
+		case <-ctx.Done():
+			return "", nil, fmt.Errorf("cancelled by context; %w", ctx.Err())
+		case <-listener.closeCh:
+			return "", nil, errors.New("closed listener")
+		case <-waitCh:
+		}
+	}
+}
+
+func (listener *Listener) set(mountPoint string, event *Event) {
+	listener.mutex.Lock()
+	defer listener.mutex.Unlock()
+
+	listener.keys = append(listener.keys, mountPoint)
+	listener.eventMap[mountPoint] = event
+
+	if listener.waitCh != nil {
+		close(listener.waitCh)
+		listener.waitCh = nil
+	}
+}
+
+func (listener *Listener) start() {
+	go func() {
+		events := make([]syscall.EpollEvent, 16)
+		defer listener.Close()
+		for {
+			select {
+			case <-listener.closeCh:
+				return
+			default:
+				if err := syscall.SetNonblock(int(listener.mountFile.Fd()), true); err != nil {
+					klog.V(3).InfoS("unable to listen for mount events", "err", err)
+					continue
+				}
+				_, err := syscall.EpollWait(listener.epollFd, events[:], resyncTimeout)
+				if err != nil {
+					klog.V(3).InfoS("unable to epoll_wait", "err", err)
+					continue
+				}
+				currentMountInfo, err := ProbeMounts()
+				if err != nil {
+					klog.V(3).InfoS("unable to probe mounts while monitoring events", "err", err)
+					continue
+				}
+				if diff := getMountDiffs(toMountMap(currentMountInfo), listener.prevMountMap); diff.mountsChanged() {
+					for k, v := range diff.toEventMap() {
+						listener.set(k, v)
+					}
+				}
```

Comment:
This is a blocking code makes blocking epoll and ctx handling.

0. Have a singleton goroutine which accepts previous and current mountinfo in a buffered channel.
1. Fetch mount information in this `for` and get previous info and set current info as previous info atomically.
2. Send previous and current info to singleton goroutine. There is no need to have a `diff` type there.
3. If mount point in current info doesn't exist in previous info, send it as Change event
4. If mount point in current info exist in previous info with interesting fields difference (not a blind equal check), send it as Change event
5. Whatever leftover mount point in previous info, send it as Delete event